File: results/0.5/Modbus/modbus-gpt-4-turbo.c
#include <hammer/hammer.h>
#include <stdio.h>
#include <stdlib.h>

// Define Modbus TCP constants
#define MODBUS_TCP_TRANSACTION_ID_SIZE 2
#define MODBUS_TCP_PROTOCOL_ID_SIZE 2
#define MODBUS_TCP_LENGTH_SIZE 2
#define MODBUS_TCP_UNIT_ID_SIZE 1
#define MODBUS_TCP_MIN_SIZE (MODBUS_TCP_TRANSACTION_ID_SIZE + MODBUS_TCP_PROTOCOL_ID_SIZE + MODBUS_TCP_LENGTH_SIZE + MODBUS_TCP_UNIT_ID_SIZE)

// Define Modbus RTU constants
#define MODBUS_RTU_ADDRESS_SIZE 1
#define MODBUS_RTU_CRC_SIZE 2

// Function codes
enum ModbusFunctionCodes {
    READ_COILS = 0x01,
    READ_DISCRETE_INPUTS = 0x02,
    READ_HOLDING_REGISTERS = 0x03,
    READ_INPUT_REGISTERS = 0x04,
    WRITE_SINGLE_COIL = 0x05,
    WRITE_SINGLE_REGISTER = 0x06,
    WRITE_MULTIPLE_COILS = 0x0F,
    WRITE_MULTIPLE_REGISTERS = 0x10
};

// Parser for a Modbus TCP frame
HParser *modbus_tcp_frame() {
    HParser *p_transaction_id = h_uint16();
    HParser *p_protocol_id = h_uint16();
    HParser *p_length = h_uint16();
    HParser *p_unit_id = h_uint8();
    HParser *p_function_code = h_uint8();
    HParser *p_data = h_bits(8, false);  // Placeholder for data parsing based on function code

    return h_sequence(p_transaction_id, p_protocol_id, p_length, p_unit_id, p_function_code, p_data, NULL);
}

// Parser for a Modbus RTU frame
HParser *modbus_rtu_frame() {
    HParser *p_address = h_uint8();
    HParser *p_function_code = h_uint8();
    HParser *p_data = h_bits(8, false);  // Placeholder for data parsing based on function code
    HParser *p_crc = h_uint16();

    return h_sequence(p_address, p_function_code, p_data, p_crc, NULL);
}

int main(int argc, char *argv[]) {
    if (argc != 2) {
        fprintf(stderr, "Usage: %s <modbus_binary_file>\n", argv[0]);
        return 1;
    }

    // Open the binary file
    FILE *file = fopen(argv[1], "rb");
    if (!file) {
        perror("Failed to open file");
        return 1;
    }

    // Get the file size
    fseek(file, 0, SEEK_END);
    long file_size = ftell(file);
    rewind(file);

    // Read the file into memory
    uint8_t *buffer = malloc(file_size);
    if (buffer == NULL) {
        fprintf(stderr, "Memory allocation failed\n");
        fclose(file);
        return 1;
    }
    fread(buffer, 1, file_size, file);

    // Parse the buffer based on Modbus TCP or RTU
    HParser *modbus_parser = (file_size > MODBUS_TCP_MIN_SIZE) ? modbus_tcp_frame() : modbus_rtu_frame();
    HParseResult *result = h_parse(modbus_parser, buffer, file_size);

    if (result == NULL) {
        fprintf(stderr, "Failed to parse Modbus data\n");
    } else {
        printf("Modbus data parsed successfully\n");
    }

    // Clean up
    h_parse_result_free(result);
    free(buffer);
    fclose(file);

    return 0;
}

File: results/0.5/Modbus/modbus-gpt-4-turbo.rs
use nom::{
    bytes::complete::take,
    number::complete::{be_u16, be_u8},
    IResult,
};
use std::env;
use std::fs::File;
use std::io::{self, Read};

#[derive(Debug)]
enum ModbusFunction {
    ReadCoils = 0x01,
    ReadDiscreteInputs = 0x02,
    ReadHoldingRegisters = 0x03,
    ReadInputRegisters = 0x04,
    WriteSingleCoil = 0x05,
    WriteSingleRegister = 0x06,
    WriteMultipleCoils = 0x0F,
    WriteMultipleRegisters = 0x10,
    // Add other function codes as needed
}

#[derive(Debug)]
struct ModbusHeader {
    transaction_id: u16,
    protocol_id: u16,
    length: u16,
    unit_id: u8,
}

#[derive(Debug)]
struct ModbusRequest {
    header: ModbusHeader,
    function_code: ModbusFunction,
    data: Vec<u8>,
}

fn parse_modbus_header(input: &[u8]) -> IResult<&[u8], ModbusHeader> {
    let (input, transaction_id) = be_u16(input)?;
    let (input, protocol_id) = be_u16(input)?;
    let (input, length) = be_u16(input)?;
    let (input, unit_id) = be_u8(input)?;

    Ok((
        input,
        ModbusHeader {
            transaction_id,
            protocol_id,
            length,
            unit_id,
        },
    ))
}

fn parse_modbus_request(input: &[u8]) -> IResult<&[u8], ModbusRequest> {
    let (input, header) = parse_modbus_header(input)?;
    let (input, function_code) = be_u8(input)?;
    let function_code = match function_code {
        0x01 => ModbusFunction::ReadCoils,
        0x02 => ModbusFunction::ReadDiscreteInputs,
        0x03 => ModbusFunction::ReadHoldingRegisters,
        0x04 => ModbusFunction::ReadInputRegisters,
        0x05 => ModbusFunction::WriteSingleCoil,
        0x06 => ModbusFunction::WriteSingleRegister,
        0x0F => ModbusFunction::WriteMultipleCoils,
        0x10 => ModbusFunction::WriteMultipleRegisters,
        _ => return Err(nom::Err::Error((input, nom::error::ErrorKind::NoneOf))),
    };

    let (input, data) = take(header.length - 2)(input)?;

    Ok((
        input,
        ModbusRequest {
            header,
            function_code,
            data: data.to_vec(),
        },
    ))
}

fn main() -> io::Result<()> {
    let args: Vec<String> = env::args().collect();
    if args.len() != 2 {
        eprintln!("Usage: {} <MODBUS_REQUEST_FILE>", args[0]);
        return Ok(());
    }

    let mut file = File::open(&args[1])?;
    let mut buffer = Vec::new();
    file.read_to_end(&mut buffer)?;

    match parse_modbus_request(&buffer) {
        Ok((_, request)) => println!("{:?}", request),
        Err(e) => println!("Failed to parse Modbus request: {:?}", e),
    }

    Ok(())
}